analyzer/src/steps/typing.rs

```diff
@@ -122,339 +172,521 @@ fn apply_types_to_source(
     }
 }
 
-/// Ascribes types to the given expression.
-///
-/// In case of an error, the expression is still returned, but the type is set to [`ERROR`].
-fn ascribe_types(
+fn ascribe_literal(force_string: bool, lit: &Literal) -> TypedExpr {
```

Comment:
The `force_string` parameter seems to be too specific. As it is the typing phase, no optimization should be done here.